atomic_reactor/plugins/pre_distribution_scope.py

```diff
@@ -0,0 +1,107 @@
+
+from atomic_reactor.plugin import PreBuildPlugin
+from dockerfile_parse import DockerfileParser
+
+
+class NothingToCheck(Exception):
+    pass
+
+
+class DisallowedDistributionScope(Exception):
+    pass
+
+
+class DistributionScopePlugin(PreBuildPlugin):
+    """Apply distribution-scope rule.
+
+    Possible values for this label are:
+
+    (most restrictive)
+    - private
+    - authoritative-source-only
+    - restricted
+    - public
+    (least restrictive)
+
+    The rule we want to apply is to prevent images from having a
+    less restrictive scope than their parent images.
+
+    If the distribution-scope for this image is set to a less
+    restrictive value than the distribution-scope label inherited from
+    the parent image, fail the build.
+    """
+
+    # Name of this plugin
+    key = 'distribution_scope'
+
+    # The label and its possible values.
+    # https://github.com/projectatomic/ContainerApplicationGenericLabels
+
+    # LABEL name used for specifying scope
+    SCOPE_LABEL = 'distribution-scope'
+
+    # Valid distribution-scope choice values, most restricted first
+    SCOPE_NAME = [
+        "private",
+        "authoritative-source-only",
+        "restricted",
+        "public",
+    ]
+
+    def __init__(self, tasker, workflow):
+        """
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        """
+
+        self.workflow = workflow
+
+    def get_scope(self, which, labels):
+        """
+        :param which: str, description of the image this belongs to
+        :param labels: dict, labels on the image
+        """
+
+        try:
+            scope_choice = labels[self.SCOPE_LABEL]
+        except KeyError:
+            self.log.debug("no distribution scope set for %s image", which)
+            raise NothingToCheck
+
+        try:
+            scope = self.SCOPE_NAME.index(scope_choice)
+        except ValueError:
+            self.log.warning("invalid label %s=%s for %s image",
+                             self.SCOPE_LABEL, scope_choice, which)
+            raise NothingToCheck
+
+        return scope
+
+    def run(self):
+        try:
+            # Find out the intended scope for this image
+            labels = DockerfileParser(self.workflow.builder.df_path).labels
+            scope = self.get_scope('current', labels)
+
+            # Find out the parent's intended scope
+            parent_labels = self.workflow.builder.base_image_inspect['Labels']
+            parent_scope = self.get_scope('parent', parent_labels)
+        except NothingToCheck:
+            self.log.debug("no checks performed")
+            return
+        else:
+            if scope > parent_scope:
+                self.log.error("%s=%s set but parent has %s=%s",
+                               self.SCOPE_LABEL, self.SCOPE_NAME[scope],
+                               self.SCOPE_LABEL, self.SCOPE_NAME[parent_scope])
+                raise DisallowedDistributionScope
```

Comment:
If we use the error message to initialize the exception, it will show up in the results section in koji making it easier to debug.